src/shogun/mathematics/graph/Storage.h

```diff
@@ -0,0 +1,395 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Gil Hoben
+ */
+
+#ifndef SHOGUN_OPERATOR_STORAGE_H_
+#define SHOGUN_OPERATOR_STORAGE_H_
+
+#include <shogun/lib/memory.h>
+#include <shogun/mathematics/graph/Shape.h>
+#include <shogun/mathematics/graph/Types.h>
+#include <shogun/util/zip_iterator.h>
+
+#include <numeric>
+#include <utility>
+
+namespace shogun
+{
+	namespace graph
+	{
+		namespace op
+		{
+			class ReshapeShogun;
+		}
+		/* A storage abstraction
+		 * The user should never see this. This is purely for
+		 * the backend to allocate storage. The user should use
+		 * the tensor class exclusively, in order to not mess
+		 * up shapes and so on.
+		 *
+		 */
+		class Storage
+		{
+			/* Managed storage. Its only job is to keep track of memory
+			 * and how to delete and allocate it.
+			 * It's unaware of what it actually owns. This is pretty much
+			 * a wrapper around std::shared_ptr for aligned memory.
+			 * Should this be thread safe?
+			 */
+			class MemoryManager
+			{
+				struct Move
+				{
+				};
+				struct NonOwning
+				{
+				};
+
+				using deleted_unique_ptr =
+				    std::unique_ptr<void, std::function<void(void*)>>;
+
+			public:
+				MemoryManager(
+				    size_t size, const std::shared_ptr<NumberType>& type)
+				{
+					void* data = sg_aligned_malloc(
+					    size_in_bytes(size, type),
+					    alignment::container_alignment);
+					m_internal_data = deleted_unique_ptr(data, [](void* ptr) {
+						if (ptr)
+							SG_ALIGNED_FREE(ptr);
+					});
+				}
+
+				MemoryManager(void* data)
+				    : m_internal_data(data, [](void* ptr) {
+					      if (ptr)
+						      SG_ALIGNED_FREE(ptr);
+				      })
+				{
+				}
+
+				/* Copies a pointer to MemoryManager
+				 */
+				static std::unique_ptr<MemoryManager> copy_from(
+				    void* ptr, size_t size,
+				    const std::shared_ptr<NumberType>& type)
+				{
+					auto result = std::make_unique<MemoryManager>(size, type);
+					sg_memcpy(
+					    result->m_internal_data.get(), ptr,
+					    size_in_bytes(size, type));
+					return result;
+				}
+
+				/* Moves ownership of the pointer to MemoryManager
+				 */
+				static std::unique_ptr<MemoryManager> move_from(
+				    void* ptr, size_t size,
+				    const std::shared_ptr<NumberType>& type)
+				{
+					return std::make_unique<MemoryManager>(
+					    ptr, size, type, Move{});
+				}
+
+				/* Generates a non owning view of the pointer
+				 */
+				static std::unique_ptr<MemoryManager> view(
+				    void* ptr, size_t size,
+				    const std::shared_ptr<NumberType>& type)
+				{
+					return std::make_unique<MemoryManager>(
+					    ptr, size, type, NonOwning{});
+				}
+
+				void
+				realloc(size_t size, const std::shared_ptr<NumberType>& type)
+				{
+					void* new_mem = std::realloc(
+					    m_internal_data.get(), size_in_bytes(size, type));
```

Comment:
```suggestion
					void* new_mem = SG_ALIGNED_REALLOC(
					    m_internal_data.get(), 0, size_in_bytes(size, type), alignment::container_alignment);
```

actually this should work